core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -51,6 +51,9 @@ class BlockManagerMasterEndpoint(
   // Mapping from block manager id to the block manager's information.
   private val blockManagerInfo = new mutable.HashMap[BlockManagerId, BlockManagerInfo]
 
+  // Mapping from executor id to the block manager's local disk directories.
+  private val executorIdToLocalDirs = new mutable.HashMap[String, Array[String]]
```

Comment:
so this grows without bound?  That's a problem for super long lived streaming apps.  I think it needs to be bounded in size, even if large.  (worst-case, you don't know the dirs and go back to fetching over the network, right?)

Reply:
Yes, it grows without limit. Right now the code counts on there is a local dir for each of the requested executor so if we would like to limit this map the fetching logic must be changed too.

But is not the same growing is present in the external shuffle service (even it could be worse there as it grows for each parallel Spark apps)?

Reply:
but this is on the driver, which needs to store the locations of all executors ever, not just the executors on that node.

though you are right, there is the same problem in principle on the external shuffle service, and the driver at least has memory dedicated to this application.  What do other reviewers think?

Reply:
@vanzin I would like to kindly ask your opinion about the unbounded `executorIdToLocalDirs`

Reply:
I don't think letting it grow unbounded is a good idea.

One idea that I haven't really thought through (or checked whether it's viable) is to try to look into the `MapOutputTracker` and clean this up when there are no more known outputs needed from that executor's host.

(Or maybe even keep this cache inside the `MapOutputTracker`?)

Otherwise, having a limit to this cache would be good (e.g. with a LRU eviction policy). Worst thing is what Imran says, you'll go over the network.